File: backend/services/cloudinaryService.js
const cloudinary = require('cloudinary').v2;

// Configure Cloudinary
cloudinary.config({
  cloud_name: process.env.CLOUDINARY_CLOUD_NAME,
  api_key: process.env.CLOUDINARY_API_KEY,
  api_secret: process.env.CLOUDINARY_API_SECRET,
});

/**
 * Upload a base64-encoded image or video to Cloudinary.
 * @param {string} fileData - base64 data URI (e.g. "data:image/png;base64,...")
 * @param {object} options
 * @param {string} options.folder - Cloudinary folder (e.g. "doctors/photos")
 * @param {'image'|'video'|'auto'} options.resourceType - Resource type
 * @param {string} [options.publicId] - Optional custom public ID
 * @returns {Promise<{url: string, publicId: string, format: string, width: number, height: number, bytes: number}>}
 */
async function uploadToCloudinary(fileData, { folder = 'arogya-raksha', resourceType = 'auto', publicId } = {}) {
  const uploadOptions = {
    folder,
    resource_type: resourceType,
    quality: 'auto',
    fetch_format: 'auto',
    timeout: 120000,        // 120 second timeout
    chunk_size: 6000000,    // 6MB chunks for large files
  };
  if (publicId) uploadOptions.public_id = publicId;

  // For videos, add eager transformations
  if (resourceType === 'video') {
    uploadOptions.eager = [
      { width: 640, crop: 'scale', format: 'mp4' },
    ];
    uploadOptions.eager_async = true;
  }

  // For images, compress before uploading to reduce transfer size
  if (resourceType === 'image') {
    uploadOptions.transformation = [
      { width: 1200, crop: 'limit' },
      { quality: 'auto' },
    ];
  }

  const result = await cloudinary.uploader.upload(fileData, uploadOptions);

  return {
    url: result.secure_url,
    publicId: result.public_id,
    format: result.format,
    width: result.width,
    height: result.height,
    bytes: result.bytes,
    duration: result.duration || null,
    resourceType: result.resource_type,
    thumbnailUrl: result.resource_type === 'video'
      ? cloudinary.url(result.public_id, { resource_type: 'video', format: 'jpg', transformation: [{ width: 400, crop: 'fill' }] })
      : null,
  };
}

/**
 * Delete a file from Cloudinary by its public ID.
 */
async function deleteFromCloudinary(publicId, resourceType = 'image') {
  const result = await cloudinary.uploader.destroy(publicId, { resource_type: resourceType });
  return result;
}

/**
 * Generate a signed upload URL for direct browser uploads.
 */
function generateUploadSignature(folder = 'arogya-raksha') {
  const timestamp = Math.round(new Date().getTime() / 1000);
  const signature = cloudinary.utils.api_sign_request(
    { timestamp, folder },
    process.env.CLOUDINARY_API_SECRET
  );
  return {
    timestamp,
    signature,
    cloudName: process.env.CLOUDINARY_CLOUD_NAME,
    apiKey: process.env.CLOUDINARY_API_KEY,
    folder,
  };
}

module.exports = {
  cloudinary,
  uploadToCloudinary,
  deleteFromCloudinary,
  generateUploadSignature,
};

File: backend/services/cloudinary_service.py
"""
cloudinary_service.py — Cloudinary upload/delete helpers.
Mirrors services/cloudinaryService.js.
"""
import hashlib
import hmac
import time
from typing import Optional

import cloudinary
import cloudinary.uploader
from core.config import settings

cloudinary.config(
    cloud_name=settings.CLOUDINARY_CLOUD_NAME,
    api_key=settings.CLOUDINARY_API_KEY,
    api_secret=settings.CLOUDINARY_API_SECRET,
)

_configured = bool(
    settings.CLOUDINARY_CLOUD_NAME
    and settings.CLOUDINARY_API_KEY
    and settings.CLOUDINARY_API_SECRET
)

if not _configured:
    print("❌ Missing Cloudinary credentials — media uploads will fail!")
else:
    print(f"☁️  Cloudinary configured for cloud: {settings.CLOUDINARY_CLOUD_NAME}")


def upload_to_cloudinary(
    file_data: str,
    folder: str = "arogya-raksha",
    resource_type: str = "auto",
    public_id: Optional[str] = None,
) -> dict:
    """
    Upload a base64 data URI or file path to Cloudinary.
    Returns { url, public_id, resource_type, format }.
    """
    if not _configured:
        raise RuntimeError("Cloudinary not configured. Missing credentials.")
    result = cloudinary.uploader.upload(
        file_data,
        folder=folder,
        resource_type=resource_type,
        **({"public_id": public_id} if public_id else {}),
    )
    return {
        "url": result.get("secure_url"),
        "public_id": result.get("public_id"),
        "resource_type": result.get("resource_type"),
        "format": result.get("format"),
        "bytes": result.get("bytes"),
        "width": result.get("width"),
        "height": result.get("height"),
    }


def delete_from_cloudinary(public_id: str, resource_type: str = "image") -> dict:
    """Delete a Cloudinary asset by public_id."""
    if not _configured:
        return {"result": "not_configured"}
    result = cloudinary.uploader.destroy(public_id, resource_type=resource_type)
    return result


def generate_upload_signature(folder: str = "arogya-raksha") -> dict:
    """Generate a signed upload preset for direct browser → Cloudinary uploads."""
    if not _configured:
        raise RuntimeError("Cloudinary not configured.")
    timestamp = int(time.time())
    params = f"folder={folder}&timestamp={timestamp}"
    signature = hmac.new(
        settings.CLOUDINARY_API_SECRET.encode(),
        params.encode(),
        hashlib.sha256,
    ).hexdigest()
    return {
        "signature": signature,
        "timestamp": timestamp,
        "cloudName": settings.CLOUDINARY_CLOUD_NAME,
        "apiKey": settings.CLOUDINARY_API_KEY,
        "folder": folder,
    }
